src/main/java/studentlist/StudentList.java

```diff
@@ -53,47 +114,160 @@ public int getNumStudents() {
+    public void printStudentList() {
+        try {
+            ui.printStudentList(this.students);
+        } catch (SEPException e) {
+            ui.printResponse(e.getMessage());
         }
+    }
 
-        // Create and return the Student object
-        return new Student(parts[1].trim(), Float.parseFloat(parts[2].trim()), preferences);
+    public void generateReport() {
+        try {
+            ui.generateReport(students);
+        } catch (SEPException e) {
+            ui.printResponse(e.getMessage());
+        }
     }
 
+    /**
+     * Updates the current student list with a new one.
+     *
+     * @param studentList The new student list to set.
+     */
+    public void setStudentList(StudentList studentList) {
+        this.students = studentList.students; // Overwrite the current ArrayList
+    }
 
-    public void addStudent(Student student) {
-        students.add(student);
+    /**
+     * Organizes the student ID by removing all spaces,
+     * including any spaces in between the numbers or alphabets,
+     * before validating the student ID.
+     *
+     * @param id The raw student ID string.
+     * @return A trimmed and space-free student ID.
+     */
+    private String organiseId(String id) {
+        String studentId = id.trim();
+        return studentId.replaceAll("\\s+", "");
     }
 
-    public void deleteStudent(String id) {
-        String[] parts = id.split(" ");
-        students.removeIf(student -> student.getId().equals(parts[1]));
+    /**
+     * Splits the input string into parts based on the expected format
+     * and order: id/, gpa/, followed by p/. Any other order would be
+     * rejected and a format exception would be thrown.
+     *
+     * @param input The raw input string.
+     * @return A String array containing the parts of the input.
+     * @throws SEPFormatException If the input format or order is invalid.
+     */
+    private String[] splitInput(String input) throws SEPException {
+        // Updated regex to allow spaces in ID, GPA, and preferences
+        String pattern = "^add\\s+id/[\\S ]+\\s+gpa/[\\S ]+\\s+p/\\{[\\S ]+}$";
+        if (!input.matches(pattern)) {
+            throw SEPFormatException.rejectAddStudentFormat();
+        }
+        String[] parts = input.split("id/|gpa/|p/");
+        if (parts.length != 4) {
+            throw SEPFormatException.rejectAddStudentFormat();
+        }
+        return parts;
     }
 
-    public void printStudentList() {
-        ui.printStudentList(this.students);
+    /**
+     * Validates the student ID format.
+     *
+     * @param studentId The student ID to validate.
+     * @param errorMessages A set to collect error messages.
+     */
+    private void validateStudentId(String studentId, Set<String> errorMessages) {
+        Pattern idPattern = Pattern.compile("^[A-Z]\\d{7}[A-Z]$");
+        if (!idPattern.matcher(studentId).matches()) {
+            errorMessages.add(SEPFormatException.rejectIdFormat().getMessage());
```

Comment:
There are multiple instances where you make consecutive function calls in 1 line. You can consider refactoring it into a variable to make it less cluttered.

Reply:
Hi, I believe this is unnecessary, considering most of these are of one time use and still readable